Tools/CISupport/ews-build/steps.py

```diff
@@ -5929,6 +5930,83 @@ def getResultSummary(self):
             configuration += ', Xcode: {}'.format(xcode_version)
         return {'step': configuration}
 
+class CheckOutLLVMProject(git.Git, AddToLogMixin):
+    name = 'checkout-llvm-project'
+    directory = 'llvm-project'
+    branch = 'webkit'
+    CHECKOUT_DELAY_AND_MAX_RETRIES_PAIR = (0, 2)
+    GIT_HASH_LENGTH = 40
+    haltOnFailure = False
+
+    def __init__(self, **kwargs):
+        repourl = f'{GITHUB_URL}/rniwa/llvm-project.git'
+        super().__init__(
+            repourl=repourl,
+            workdir=self.directory,
+            retry=self.CHECKOUT_DELAY_AND_MAX_RETRIES_PAIR,
+            timeout=2 * 60 * 60,
+            alwaysUseLatest=False,
+            logEnviron=False,
+            progress=True,
+            **kwargs
+        )
+
+    @defer.inlineCallbacks
+    def run_vc(self, branch, revision, patch):
+        rc = yield super().run_vc(self.branch, CLANG_VERSION, None)
+        return rc
+
+    @defer.inlineCallbacks
+    def parseGotRevision(self, _=None):
+        stdout = yield self._dovccmd(['rev-parse', 'HEAD'], collectStdout=True)
+        revision = stdout.strip()
+        if len(revision) != self.GIT_HASH_LENGTH:
+            raise buildstep.BuildStepFailed()
+        return SUCCESS
+
+    def getResultSummary(self):
+        if self.results != SUCCESS:
+            return {'step': 'Failed to update llvm-project directory'}
+        else:
+            return {'step': 'Cleaned and updated llvm-project directory'}
+
+
+class UpdateClang(steps.ShellSequence, ShellMixin):
+    name = 'update-clang'
+    description = 'updating clang'
+    descriptionDone = 'updated clang'
+    flunkOnFailure = False
+    warnOnFailure = False
+
+    def __init__(self, **kwargs):
+        super().__init__(logEnviron=False, workdir=LLVM_DIR, **kwargs)
+        self.commands = []
+
+    @defer.inlineCallbacks
+    def run(self):
+        # ['xcrun', 'cmake', '-DLLVM_ENABLE_PROJECTS=clang', '-DCMAKE_BUILD_TYPE=Release', '-G', 'Ninja', '../llvm', '-DCMAKE_MAKE_PROGRAM=$(xcrun --sdk macosx --find ninja)']
+        command_list = [
+            ['rm', '-r', 'build'],
+            ['mkdir', 'build'],
+            ['git', 'apply', 'center-endpath-html.patch'],
+            ['cd', 'build'],
+            # TODO: these are not running??
+            self.shell_command('xcrun cmake -DLLVM_ENABLE_PROJECTS=clang -DCMAKE_BUILD_TYPE=Release -G Ninja ../llvm -DCMAKE_MAKE_PROGRAM=$(xcrun --sdk macosx --find ninja)'),
+            self.shell_command('ninja clang')
+        ]
+        for command in command_list:
+            self.commands.append(util.ShellArg(command=command, logname='stdio'))
+        rc = yield super().run()
+        defer.returnValue(rc)
+
+    def doStepIf(self, step):
+        return self.build.getProperty('llvm_revision', '') != CLANG_VERSION
```

Comment:
Logic looks fine.
Also, consider what behavior do you want if llvm_revision is not defined for some reason (e.g.: previous step failed to extract the revision info). Currently, it would skip the update in that case (which looks reasonable).

Nit: might be good to have more consistency in the variable naming (`llvm_revision` being compared with `CLANG_VERSION`)

Reply:
Changed variables to be `llvm_revision`.

If `llvm_revision` is not defined, we should actually update. For example, the first time we bring up a new machine would result in `llvm_revision == ''` and we want to checkout and build.